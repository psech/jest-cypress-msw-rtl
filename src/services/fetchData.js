import axios from 'axios';

export const BASE_URL = 'https://jsonplaceholder.typicode.com';

export const fetchData = async () => {
  try {
    const response = await axios.get(`${BASE_URL}/todos`);
    if (response.status === 200) {
      return response.data.slice(0, 10);
    }
  } catch (error) {
    console.error(error);
    return [];
  }
};
