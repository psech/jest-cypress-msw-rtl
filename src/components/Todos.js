import './Todos.css';
import React, { useState, useEffect } from 'react';
import { fetchData } from '../services/fetchData.js';
import Todo from './Todo';

export default function Todos() {
  const [todos, setTodos] = useState([]);
  const [error, setError] = useState({ isError: false, message: null });

  const getSetData = async () => {
    const data = await fetchData();
    setTodos(data);

    if (data.length === 0) {
      setError({ isError: true, message: 'Something went wrong. 🤷🏻‍♂️' });
    }
  };

  useEffect(() => {
    getSetData();
  }, []);

  const handleChange = (id) =>
    setTodos((prevTodos) =>
      prevTodos.map((todo) =>
        todo.id === id ? { ...todo, completed: !todo.completed } : todo
      )
    );

  const handleRefresh = () => window.location.reload(false);

  return (
    <>
      <h3>Todos</h3>
      <button onClick={handleRefresh}>Refresh</button>
      <table data-testid="todo-table">
        <thead>
          <tr>
            <th>#</th>
            <th>Todo</th>
            <th>Completed</th>
          </tr>
        </thead>
        <tbody>
          {error.isError ? (
            <tr>
              <td colspan="3" style={{ color: 'red' }}>
                {error.message}
              </td>
            </tr>
          ) : (
            todos.map((todo) => (
              <Todo key={todo.id} todo={todo} onChange={handleChange} />
            ))
          )}
        </tbody>
      </table>
    </>
  );
}
