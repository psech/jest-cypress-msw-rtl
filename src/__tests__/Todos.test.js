import React from 'react';
import { render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import Todos from '../components/Todos';
import axios from 'axios';
import * as service from '../services/fetchData.js';

jest.mock('axios');

describe('Todos', () => {
  beforeEach(() => {
    jest.spyOn(React, 'useEffect').mockImplementation((f) => f());
  });

  test('Present data', () => {
    const mockObj = (idx) => ({
      userId: idx,
      id: idx,
      title: 'delectus aut autem',
      completed: false,
    });

    const expectedTodos = [...Array(2).keys()].map((i) => mockObj(i + 1));
    const mockData = {
      data: expectedTodos,
      status: 200,
    };
    axios.get.mockResolvedValueOnce(mockData);

    jest.spyOn(service, 'fetchData');

    act(() => {
      render(<Todos />);
    });

    const table = screen.getByTestId('todo-table');
    expect(table).toMatchSnapshot();

    expect(service.fetchData).toHaveBeenCalledTimes(1);
  });
});
