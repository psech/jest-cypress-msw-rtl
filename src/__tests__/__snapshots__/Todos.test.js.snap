// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos Present data 1`] = `
<table
  data-testid="todo-table"
>
  <thead>
    <tr>
      <th>
        #
      </th>
      <th>
        Todo
      </th>
      <th>
        Completed
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
